### DISCUSSION TITLE
Hyper test suite compatibility

### DISCUSSION THREAD

**Discussion by kevinburke1** [curl/curl#6835] [General]
I'd like to try to work on making the Hyper backend pass the curl test suite, which it currently doesn't.

Generally, I imagine a lot of the fixes will be something like:

1. Expose some new functionality in capi/include/hyper.h in the Hyper project.
2. Rely on that new functionality in c-hyper.c in the Curl project
3. Pass a new test case.

Is there a minimum version of Hyper that Curl depends on right now? It would be very slow if I had to add changes to hyper.h, and then wait for a new release of Hyper before merging the fixes upstream to Curl. Do you have thoughts on how to approach this?

**Comment by bagder** [comment]
Hi @kevinburkemeter, I'm happy to see you want to help out!

I think you'll notice that very little of the remaining flaws in the Hyper integration is fixed by changing Hyper itself. Most of what needs to be done is about making sure the Hyper backend behaves properly like the built-in code and uses Hyper correctly. Hyper itself probably already does the right thing, we just need to use it and tickle it the right way and in the right sequence. This is not always an easy task.

If you end up with something needing Hyper to change or even requiring a recent version of Hyper to work, then I think we just bump the lowest required version to that particular version that has the stuff fixed. We have notthing to gain by trying to support old Hyper versions until curl runs all test cases fine. Until we reach that point, I will consider the Hyper backend as "under development" and then I believe it is perfectly fine to require the bleeding edge from a dependency like this.

**Comment by kevinburke1** [comment]
Thanks! That's helpful. I guess the first test case that I noticed that was failing was, Curl returns a specific error code if server returns "HTTP/1.2", but Hyper doesn't expose the error to C in enough detail to catch that particular case. So I was going to try to add a new error constant to the Hyper FFI bindings and assumed there would be more issues like that.

**Comment by kevinburke** [comment]
Just going to add notes as I come across them.

Test 158 checks that a 100 Continue with no followup returns CURLE_GOT_NOTHING. Hyper does not support returning a 100 Continue as a Response per https://github.com/hyperium/hyper/blob/master/src/proto/h1/role.rs#L304-L307. So by the time c-hyper.c sees the response code in Curl_hyper_stream it's a 200, not a 100. 

In other words I don't know that there's a way to solve this right now.

**Reply by bagder** [reply]
This surprises me and seems like a rather big oversight. 1xx responses are important in HTTP. Do you know if there's a Hyper issue already filed for this? Do you know if there's any discussion there on fixing this?

/cc @seanmonstar

**Reply by kevinburke** [reply]
It's very possible I'm reading the code incorrectly - I still don't understand Rust very well at all. I guess we could check whether the other 100 Continue tests in curl pass or fail.

**Reply by seanmonstar** [reply]
The linked code is about the *server* API, since it's basically `fn(Request) -> Response`, and if the returned response is only a 1xx, we'd never know the final response. There's a proposal to adjust that, but I think it's off-topic for here.

hyper's client will currently handle receiving 1xx responses, but there's also a lack of an API to view them, so it just ignores them at the moment. I expect that the unit test would correctly fail, since the test server would never have sent a 200, and so hyper would know the closed connection would mean an error.

**Reply by kevinburke** [reply]
Aha, sorry for the mixup, I didn't realize Hyper had a server component as well. I'll read the code more carefully going forward.

**Comment by kevinburke** [comment]
Test 214 *was* failing but was literally fixed yesterday by amending the Rust http package to accept `{` and `}`. You need to rebuild hyper with `http` version 0.2.4 (`cargo update -p http`).

**Reply by kevinburke** [reply]
That said, I think we are returning CURLE_OUT_OF_MEMORY from c-hyper.c if you pass an invalid path, which seems incorrect. Perhaps another test will expose this.

**Reply by bagder** [reply]
What actually is "an invalid path" ?

**Reply by kevinburke** [reply]
The path parsing logic is here https://github.com/hyperium/http/blob/master/src/uri/path.rs#L21. By "an invalid path" I mean any error returned by that function instead of a path result.

**Reply by bagder** [reply]
I really wished Hyper wouldn't care what we put there... Also, that code refers to the WHATWG URL spec which is totally the wrong spec to talk about when we speak HTTP!

I don't understand the rust code. What doesn't it allow to be there?

[RFC 7230 section 5.3](https://tools.ietf.org/html/rfc7230#section-5.3) does not mention many restrictions...

**Reply by jsha** [reply]
Reading through the specs... It seems like the relevant bit is https://tools.ietf.org/html/rfc3986#section-2.2:

>    URI producing applications should percent-encode data octets that
   correspond to characters in the reserved set unless these characters
   are specifically allowed by the URI scheme to represent data in that
   component.  If a reserved character is found in a URI component and
   no delimiting role is known for that character, then it must be
   interpreted as representing the data octet corresponding to that
   character's encoding in US-ASCII.

In other words, if you're parsing a URL and encounter a reserved character, even if it's not escaped, you should just treat it as itself (according to RFC 3986). Am I interpreting correctly?
   
> I don't understand the rust code. What doesn't it allow to be there?

It allows the WHATWG's "[URL code points](https://url.spec.whatwg.org/#url-code-points)":

> The URL code points are ASCII alphanumeric, U+0021 (!), U+0024 ($), U+0026 (&), U+0027 ('), U+0028 LEFT PARENTHESIS, U+0029 RIGHT PARENTHESIS, U+002A (*), U+002B (+), U+002C (,), U+002D (-), U+002E (.), U+002F (/), U+003A (:), U+003B (;), U+003D (=), U+003F (?), U+0040 (@), U+005F (_), U+007E (~), and code points in the range U+00A0 to U+10FFFD, inclusive, excluding surrogates and noncharacters. 

Plus `?#"{}`. Other characters are rejected.

**Reply by bagder** [reply]
> f you're parsing a URL and encounter a reserved character, even if it's not escaped, you should just treat it as itself (according to RFC 3986). Am I interpreting correctly?

That's how I read it. I believe that's also how many (URL parsing) HTTP clients behave. They allow for example "{}" to appear in URLs.

I expect the URL parser to get to deal with what's okay or not in the URL and I expect the HTTP layer to then just deal with whatever octets the upper layer wants to send. I don't see the point in the HTTP layer enforcing its own set of rules unless it really has to (like bytes that aren't compatible with the wire protocol etc).

Further: leaning against a WHATWG spec is really asking for problems for a library since that spec is constantly moving so whatever you thought was correct last year is no longer right today. I want my libraries more stable than that and we need to provide a more solid answers than so to our users. HTTP/1 and HTTP/2 are specified with dependencies on RFC 3986, *not* on the WHATWG URL spec.

The differences and disagreements between the WHATWG URL spec and RFC 3986 can then be left for the URL parser to make decisions about, not the HTTP layer code.

**Reply by jsha** [reply]
That makes sense regarding the WHATWG spec.

> The differences and disagreements between the WHATWG URL spec and RFC 3986 can then be left for the URL parser to make decisions about, not the HTTP layer code.

And in this specific case, you're saying that curl has a URL parser, and by the time curl passes a request path to Hyper, it's already been through curl's parser, so Hyper should act as an HTTP layer (not an HTTP + URL layer) and pass it along unchanged? Or another way of saying it would be that we shouldn't have two different URL parsers acting on the same request?

EDIT s/should/shouldn't/

**Reply by bagder** [reply]
Yes, I think that's a pretty accurate way of describing the current situation.

**Comment by kevinburke** [comment]
Test 207 is supposed to return error 18 - CURLE_PARTIAL_FILE but instead returns error 56 - CURLE_RECV_ERROR.

https://gist.github.com/kevinburke/ab2078b78725929926ba1d1646d86a8e

**Reply by bagder** [reply]
This seems like a repeated pattern. hyper is not overly generous in describing *why* it thinks the data is wrong, and without that we leave the users wondering. It would be good to have hyper be more specific here so that we can return the same error for that case - and possibly even with details in failf().

**Comment by kevinburke** [comment]
Test 217 fails, because data in the body of a 405 in a proxy CONNECT request is expected to be ignored / not returned, but Hyper or c-hyper is returning it.

https://gist.github.com/kevinburke/6dcb7475e6c66dd1ac18cf0e556ba0a7

**Reply by kevinburke** [reply]
Reading through the RFC, it doesn't seem like it's either required for clients to ignore the response body, or not, so I'm not sure that either curl or Hyper's behavior is correct. It's possible we should implement the behavior in c-hyper.c to make the clients line up. https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.6

**Reply by bagder** [reply]
Thanks for the details. I'll also dig a little on this.

**Reply by bagder** [reply]
I think in the built-in CONNECT handler case, it is more a case of aborting sooner rather than later when it can't continue anyway so it doesn't bother to read the body. With hyper we can't as easily stop reading after just the headers so I figure ignoring the response body there on 4xx CONNECT responses to match behavior is probably the way forward.

**Reply by kevinburke** [reply]
I took a look at this but didn't see an easy fix, but part of it is I've never used connectdata or Curl_easy before and need to see what fields are available.

**Comment by kevinburke** [comment]
Test 221, 222, 223, 224, 230, 232 fail on my machine because they're expecting both CR and LF to be written after an Accept-Encoding header but there's only a LF. I believe this is due to the `<strippart>` implementation matching the carriage return more than the underlying code being incorrect.

```
 221: protocol FAILED:
--- log/check-expected	2021-05-29 21:03:54.000000000 -0700
+++ log/check-generated	2021-05-29 21:03:54.000000000 -0700
@@ -2,5 +2,5 @@
 Host: 127.0.0.1:61827[CR][LF]
 User-Agent: curl/7.77.1-DEV[CR][LF]
 Accept: */*[CR][LF]
-Accept-Encoding: xxx[CR][LF]
+Accept-Encoding: xxx[LF]
 [CR][LF]
```

**Reply by kevinburke** [reply]
Maybe this is a Mac-specific issue, based on how `strippart` is implemented? I have GNU sed listed first in my $PATH, if that makes a difference.

```
# Verify data after the test has been "shot"
<verify>
<strippart>
s/^Accept-Encoding: .*/Accept-Encoding: xxx/
</strippart>
<protocol>
GET /%TESTNUMBER HTTP/1.1
Host: %HOSTIP:%HTTPPORT
User-Agent: curl/%VERSION
Accept: */*
Accept-Encoding: xxx

</protocol>
<errorcode>
61
</errorcode>
</verify>
```

**Reply by kevinburke** [reply]
I'm not super familiar with perl, but I think the issue is that ".*" is matching and then stripping the trailing carriage return. I'm able to make the test pass by amending the logic in test222 to read:

```
<strippart>
s/^Accept-Encoding: [a-zA-Z, ]*/Accept-Encoding: xxx/a
</strippart>
```

**Reply by kevinburke** [reply]
This should be fixed by https://github.com/curl/curl/pull/7171

**Comment by kevinburke** [comment]
Test 262 writes the null byte \x00 at the end of each header line before the carriage return. Curl expects to be able to read the response anyway, but Hyper fails with:

```
curl: (1) Hyper: [6] invalid HTTP status-code parsed
```

**Reply by kevinburke** [reply]
Here's the status line parsing logic that Hyper uses, I think: https://github.com/seanmonstar/httparse/blob/master/src/lib.rs#L547

```rust
fn parse_reason<'a>(bytes: &mut Bytes<'a>) -> Result<&'a str> {
    let mut seen_obs_text = false;
    loop {
        let b = next!(bytes);
        if b == b'\r' {
            expect!(bytes.next() == b'\n' => Err(Error::Status));
            return Ok(Status::Complete(unsafe {
                let bytes = bytes.slice_skip(2);
                if !seen_obs_text {
                    // all bytes up till `i` must have been HTAB / SP / VCHAR
                    str::from_utf8_unchecked(bytes)
                } else {
                    // obs-text characters were found, so return the fallback empty string
                    ""
                }
            }));
        } else if b == b'\n' {
            return Ok(Status::Complete(unsafe {
                let bytes = bytes.slice_skip(1);
                if !seen_obs_text {
                    // all bytes up till `i` must have been HTAB / SP / VCHAR
                    str::from_utf8_unchecked(bytes)
                } else {
                    // obs-text characters were found, so return the fallback empty string
                    ""
                }
            }));
        } else if !(b == 0x09 || b == b' ' || (b >= 0x21 && b <= 0x7E) || b >= 0x80) {
            return Err(Error::Status);
        } else if b >= 0x80 {
            seen_obs_text = true;
        }
    }
}
```

**Reply by kevinburke** [reply]
Reading through RFC 7230, here's the description of a valid status line:

   > The first line of a response message is the status-line, consisting
   of the protocol version, a space (SP), the status code, another
   space, a possibly empty textual phrase describing the status code,
   and ending with CRLF.

     status-line = HTTP-version SP status-code SP reason-phrase CRLF

Where reason-phrase is described as:

         reason-phrase  = *( HTAB / SP / VCHAR / obs-text )

The definitions of HTAB, SP, and VCHAR are as follows:

> The following core rules are included by reference, as defined in
   [RFC5234], Appendix B.1: ALPHA (letters), CR (carriage return), CRLF
   (CR LF), CTL (controls), DIGIT (decimal 0-9), DQUOTE (double quote),
   HEXDIG (hexadecimal 0-9/A-F/a-f), HTAB (horizontal tab), LF (line
   feed), OCTET (any 8-bit sequence of data), SP (space), and VCHAR (any
   visible [USASCII] character).

So it seems like Hyper/httparse is doing the correct thing to reject a server response with null bytes. I'm not sure of the best way to go forward.

https://datatracker.ietf.org/doc/html/rfc7230#section-3.1.2

**Reply by bagder** [reply]
Related [current discussion](https://github.com/httpwg/http2-spec/pull/846) about it in the ongoing HTTP revision/cleanup explicitly says that NUL cannot appear anywhere in the value, so the test should be updated.

**Reply by kevinburke1** [reply]
> the test should be updated.

Can you describe what you have in mind here? Replace the null byte with a different unusual character in the ASCII range?

**Reply by bagder** [reply]
We should probably remove the zero byte from the status-line and use 0x01 on the other lines perhaps...

**Comment by kevinburke** [comment]
Test 266 writes a trailer that Hyper does not include as part of the response (the "chunky-trailer: header data" line).

```
=== Start of file check-expected
 HTTP/1.1 200 funky chunky![CR][LF]
 Server: fakeit/0.9 fakeitbad/1.0[CR][LF]
 Transfer-Encoding: chunked[CR][LF]
 Trailer: chunky-trailer[CR][LF]
 Connection: mooo[CR][LF]
 [CR][LF]
 aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbcccccccccccccccccccccccccccccccc[LF]
 chunky-trailer: header data[CR][LF]
=== End of file check-expected
=== Start of file check-generated
 HTTP/1.1 200 funky chunky![CR][LF]
 Server: fakeit/0.9 fakeitbad/1.0[CR][LF]
 Transfer-Encoding: chunked[CR][LF]
 Trailer: chunky-trailer[CR][LF]
 Connection: mooo[CR][LF]
 [CR][LF]
 aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbcccccccccccccccccccccccccccccccc[LF]
=== End of file check-generated
```

**Comment by bagder** [comment]
## Test 154/155/158

These tests fail because I cannot prevent hyper from sending the request body even though the test uses `Expect: 100-continue` so it shouldn't send the body as the sever responds immediately with a 401. It makes me curious if either hyper needs to change or we need to use hyper differently to support `Expect`...

**Comment by bagder** [comment]
# test 269

This uses `--ignore-content-length` (`CURLOPT_IGNORE_CONTENT_LENGTH`) but hyper doesn't seem to have the ability to do this. The reason curl has this is because there was a time when bad headers happened. Usually they were negative due to file size wraps etc.

I will mark this test as not working for hyper and document this limitation.
